src/Stratis.Bitcoin.Features.Consensus/ProvenBlockHeaders/ProvenBlockHeaderRepository.cs

```diff
@@ -0,0 +1,378 @@
+    /// </summary>
+    public class ProvenBlockHeaderRepository : IProvenBlockHeaderRepository
+    {
+        /// <summary>Instance logger.</summary>
+        private readonly ILogger logger;
+
+        /// <summary>Access to DBreeze database.</summary>
+        private readonly DBreezeEngine dbreeze;
+
+        /// <summary>Specification of the network the node runs on - RegTest/TestNet/MainNet.</summary>
+        private readonly Network network;
+
+        /// <summary>Database key under which the <see cref="ProvenBlockHeader"/> item is stored.</summary>
+        private static readonly byte[] provenBlockHeaderKey = new byte[0];
+
+        /// <summary>Database key under which the block hash and height of a <see cref="ProvenBlockHeader"/> tip is stored.</summary>
+        private static readonly byte[] blockHashHeightKey = new byte[0];
+
+        /// <summary>DBreeze table names.</summary>
+        private const string ProvenBlockHeaderTable = "ProvenBlockHeader";
+        private const string BlockHashHeightTable = "BlockHashHeight";
+
+        /// <summary>Height of the block which is currently the tip of the <see cref="ProvenBlockHeader"/>.</summary>
+        private HashHeightPair blockHashHeightPair;
+
+        /// <summary>Current <see cref="ProvenBlockHeader"/> tip.</summary>
+        private ProvenBlockHeader provenBlockHeaderTip;
+        
+        /// <summary>
+        /// Initializes a new instance of the object.
+        /// </summary>
+        /// <param name="network">Specification of the network the node runs on - RegTest/TestNet/MainNet.</param>
+        /// <param name="folder"><see cref="ProvenBlockHeaderStore"/> folder path to the DBreeze database files.</param>
+        /// <param name="dateTimeProvider">Provider of time functions.</param>
+        /// <param name="loggerFactory">Factory to create a logger for this type.</param>
+        public ProvenBlockHeaderRepository(Network network, string folder, IDateTimeProvider dateTimeProvider, ILoggerFactory loggerFactory)
+        {
+            Guard.NotNull(network, nameof(network));
+            Guard.NotNull(folder, nameof(folder));
+
+            this.logger = loggerFactory.CreateLogger(this.GetType().FullName);
+
+            // Create the ProvenBlockHeaderStore if it doesn't exist.
+            Directory.CreateDirectory(folder);
+
+            this.dbreeze = new DBreezeEngine(folder);
+            this.network = network;
+        }
+
+        /// <inheritdoc />
+        public Task InitializeAsync()
+        {
+            Task task = Task.Run(() =>
+            {
+                this.logger.LogInformation("Initializing {0}.", nameof(ProvenBlockHeaderRepository));
+
+                Block genesis = this.network.GetGenesis();
+
+                using (DBreeze.Transactions.Transaction transaction = this.dbreeze.GetTransaction())
+                {
+                    if (this.GetTipHashHeight(transaction) == null)
```

Comment:
probably in case it's null it's better to save the tip to a local variable. 
otherwise every time we need the tip we will ask db. 

same for get tip method- it should return local variable. 

the tip from local variable should be changed when the db representation is changed.

Reply:
Changed line 71 to now check if this.blockHashHeightPair is null.

In get tip this checks this.blockHashHeightPair before getting the value from disk.